File: src/ads1292_studio/quality_gate.py
from __future__ import annotations

from dataclasses import asdict, dataclass
import json
import math
from pathlib import Path

from ads1292_studio.quality import QualityMetrics


@dataclass(frozen=True)
class QualityGate:
    min_duration_seconds: float = 8.0
    min_contact_ok_percent: float = 95.0
    min_r_peaks: int = 5
    min_hr_bpm: float = 35.0
    max_hr_bpm: float = 180.0
    require_qrs_clear: bool = True
    max_baseline_drift_counts: float | None = None
    max_noise_rms_counts: float | None = None
    max_peak_to_peak_counts: float | None = None

    def normalized(self) -> "QualityGate":
        return QualityGate(
            min_duration_seconds=max(0.0, float(self.min_duration_seconds)),
            min_contact_ok_percent=max(0.0, min(100.0, float(self.min_contact_ok_percent))),
            min_r_peaks=max(0, int(self.min_r_peaks)),
            min_hr_bpm=max(0.0, float(self.min_hr_bpm)),
            max_hr_bpm=max(0.0, float(self.max_hr_bpm)),
            require_qrs_clear=bool(self.require_qrs_clear),
            max_baseline_drift_counts=_optional_nonnegative(self.max_baseline_drift_counts),
            max_noise_rms_counts=_optional_nonnegative(self.max_noise_rms_counts),
            max_peak_to_peak_counts=_optional_nonnegative(self.max_peak_to_peak_counts),
        )


@dataclass(frozen=True)
class QualityGateResult:
    passed: bool
    failures: tuple[str, ...]

    @property
    def label(self) -> str:
        return "Pass" if self.passed else "Fail"


def evaluate_quality_gate(metrics: QualityMetrics, gate: QualityGate | None = None) -> QualityGateResult:
    gate = (gate or QualityGate()).normalized()
    failures: list[str] = []
    # A non-finite metric (NaN/inf) makes every threshold comparison False and
    # would silently pass corrupt data — fail explicitly instead.
    for name, value in (
        ("duration", metrics.duration_seconds),
        ("contact %", metrics.contact_ok_percent),
        ("median HR", metrics.hr_median_bpm),
        ("baseline drift", metrics.baseline_drift_counts),
        ("noise RMS", metrics.noise_rms_counts),
        ("peak-to-peak", metrics.peak_to_peak_counts),
    ):
        if not math.isfinite(value):
            failures.append(f"{name} is not finite ({value})")
    if metrics.duration_seconds < gate.min_duration_seconds:
        failures.append(f"duration {metrics.duration_seconds:.2f}s < {gate.min_duration_seconds:.2f}s")
    if metrics.contact_ok_percent < gate.min_contact_ok_percent:
        failures.append(f"contact {metrics.contact_ok_percent:.2f}% < {gate.min_contact_ok_percent:.2f}%")
    if gate.require_qrs_clear and not metrics.qrs_clear:
        failures.append("QRS not clear")
    if metrics.r_peaks < gate.min_r_peaks:
        failures.append(f"R peaks {metrics.r_peaks} < {gate.min_r_peaks}")
    if metrics.hr_median_bpm > 0 and metrics.hr_median_bpm < gate.min_hr_bpm:
        failures.append(f"median HR {metrics.hr_median_bpm:.1f} bpm < {gate.min_hr_bpm:.1f} bpm")
    if metrics.hr_median_bpm > gate.max_hr_bpm:
        failures.append(f"median HR {metrics.hr_median_bpm:.1f} bpm > {gate.max_hr_bpm:.1f} bpm")
    if gate.max_baseline_drift_counts is not None and metrics.baseline_drift_counts > gate.max_baseline_drift_counts:
        failures.append(
            f"baseline drift {metrics.baseline_drift_counts:.1f} counts > {gate.max_baseline_drift_counts:.1f} counts"
        )
    if gate.max_noise_rms_counts is not None and metrics.noise_rms_counts > gate.max_noise_rms_counts:
        failures.append(f"noise RMS {metrics.noise_rms_counts:.1f} counts > {gate.max_noise_rms_counts:.1f} counts")
    if gate.max_peak_to_peak_counts is not None and metrics.peak_to_peak_counts > gate.max_peak_to_peak_counts:
        failures.append(
            f"peak-to-peak {metrics.peak_to_peak_counts:.1f} counts > {gate.max_peak_to_peak_counts:.1f} counts"
        )
    return QualityGateResult(passed=not failures, failures=tuple(failures))


def quality_gate_template() -> QualityGate:
    return QualityGate()


def read_quality_gate_json(path: Path | str) -> QualityGate:
    data = json.loads(Path(path).read_text())
    return QualityGate(
        min_duration_seconds=float(data.get("min_duration_seconds", 8.0)),
        min_contact_ok_percent=float(data.get("min_contact_ok_percent", 95.0)),
        min_r_peaks=int(data.get("min_r_peaks", 5)),
        min_hr_bpm=float(data.get("min_hr_bpm", 35.0)),
        max_hr_bpm=float(data.get("max_hr_bpm", 180.0)),
        require_qrs_clear=bool(data.get("require_qrs_clear", True)),
        max_baseline_drift_counts=_json_optional_float(data.get("max_baseline_drift_counts")),
        max_noise_rms_counts=_json_optional_float(data.get("max_noise_rms_counts")),
        max_peak_to_peak_counts=_json_optional_float(data.get("max_peak_to_peak_counts")),
    ).normalized()


def write_quality_gate_json(path: Path | str, gate: QualityGate) -> None:
    output = Path(path)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(asdict(gate.normalized()), indent=2) + "\n")


def _json_optional_float(value) -> float | None:
    if value in (None, ""):
        return None
    return float(value)


def _optional_nonnegative(value: float | None) -> float | None:
    if value is None:
        return None
    return max(0.0, float(value))

File: core/src/dsp/QualityGate.cpp
// core/src/dsp/QualityGate.cpp
// Quality-gate thresholds and evaluation.
// Pure C++17, no Qt, no OS. All values double.

#include "ads1292/dsp/QualityGate.h"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <sstream>
#include <string>

namespace ads1292::dsp {

// ---------------------------------------------------------------------------
// File-local formatting helpers — replicate Python's :.2f / :.1f exactly.
// ---------------------------------------------------------------------------
namespace {

std::string f2(double v) {
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(2) << v;
    return oss.str();
}

std::string f1(double v) {
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(1) << v;
    return oss.str();
}

} // namespace

// ---------------------------------------------------------------------------
// normalized
// ---------------------------------------------------------------------------
QualityGate normalized(const QualityGate& g) {
    QualityGate n = g;
    n.min_duration_seconds   = std::max(0.0, g.min_duration_seconds);
    n.min_contact_ok_percent = std::max(0.0, std::min(100.0, g.min_contact_ok_percent));
    n.min_r_peaks            = std::max(0, g.min_r_peaks);
    n.min_hr_bpm             = std::max(0.0, g.min_hr_bpm);
    n.max_hr_bpm             = std::max(0.0, g.max_hr_bpm);
    // require_qrs_clear: keep as-is

    if (g.max_baseline_drift_counts.has_value())
        n.max_baseline_drift_counts = std::max(0.0, *g.max_baseline_drift_counts);
    else
        n.max_baseline_drift_counts = std::nullopt;

    if (g.max_noise_rms_counts.has_value())
        n.max_noise_rms_counts = std::max(0.0, *g.max_noise_rms_counts);
    else
        n.max_noise_rms_counts = std::nullopt;

    if (g.max_peak_to_peak_counts.has_value())
        n.max_peak_to_peak_counts = std::max(0.0, *g.max_peak_to_peak_counts);
    else
        n.max_peak_to_peak_counts = std::nullopt;

    return n;
}

// ---------------------------------------------------------------------------
// evaluate_quality_gate
// ---------------------------------------------------------------------------
QualityGateResult evaluate_quality_gate(
    const QualityMetrics& m,
    const QualityGate&    gate)
{
    const QualityGate g = normalized(gate);
    std::vector<std::string> failures;

    // --- Finiteness checks (in the prescribed order) ---
    struct FiniteCheck { const char* name; double value; };
    const FiniteCheck finite_checks[] = {
        { "duration",       m.duration_seconds      },
        { "contact %",      m.contact_ok_percent     },
        { "median HR",      m.hr_median_bpm          },
        { "baseline drift", m.baseline_drift_counts  },
        { "noise RMS",      m.noise_rms_counts       },
        { "peak-to-peak",   m.peak_to_peak_counts    },
    };
    for (const auto& fc : finite_checks) {
        if (!std::isfinite(fc.value)) {
            // Format the non-finite value the same way Python does:
            // float('nan') → "nan", float('inf') → "inf"
            std::string val_str;
            if (std::isnan(fc.value))
                val_str = "nan";
            else if (fc.value > 0)
                val_str = "inf";
            else
                val_str = "-inf";
            failures.push_back(
                std::string(fc.name) + " is not finite (" + val_str + ")");
        }
    }

    // --- Threshold checks (EXACT failure strings, matching quality_gate.py) ---

    if (m.duration_seconds < g.min_duration_seconds)
        failures.push_back(
            "duration " + f2(m.duration_seconds) + "s < " + f2(g.min_duration_seconds) + "s");

    if (m.contact_ok_percent < g.min_contact_ok_percent)
        failures.push_back(
            "contact " + f2(m.contact_ok_percent) + "% < " + f2(g.min_contact_ok_percent) + "%");

    if (g.require_qrs_clear && !m.qrs_clear)
        failures.push_back("QRS not clear");

    if (m.r_peaks < g.min_r_peaks)
        failures.push_back(
            "R peaks " + std::to_string(m.r_peaks) + " < " + std::to_string(g.min_r_peaks));

    if (m.hr_median_bpm > 0.0 && m.hr_median_bpm < g.min_hr_bpm)
        failures.push_back(
            "median HR " + f1(m.hr_median_bpm) + " bpm < " + f1(g.min_hr_bpm) + " bpm");

    if (m.hr_median_bpm > g.max_hr_bpm)
        failures.push_back(
            "median HR " + f1(m.hr_median_bpm) + " bpm > " + f1(g.max_hr_bpm) + " bpm");

    if (g.max_baseline_drift_counts.has_value() &&
        m.baseline_drift_counts > *g.max_baseline_drift_counts)
        failures.push_back(
            "baseline drift " + f1(m.baseline_drift_counts) +
            " counts > " + f1(*g.max_baseline_drift_counts) + " counts");

    if (g.max_noise_rms_counts.has_value() &&
        m.noise_rms_counts > *g.max_noise_rms_counts)
        failures.push_back(
            "noise RMS " + f1(m.noise_rms_counts) +
            " counts > " + f1(*g.max_noise_rms_counts) + " counts");

    if (g.max_peak_to_peak_counts.has_value() &&
        m.peak_to_peak_counts > *g.max_peak_to_peak_counts)
        failures.push_back(
            "peak-to-peak " + f1(m.peak_to_peak_counts) +
            " counts > " + f1(*g.max_peak_to_peak_counts) + " counts");

    return { failures.empty(), failures };
}

} // namespace ads1292::dsp
